File: ios/Betty/Features/Live/PushRegistrationService.swift
import Foundation
import FirebaseCore
import FirebaseMessaging
import Observation
import UIKit
import UserNotifications

/// Firebase Messaging registration + `POST /user/me/add_push_token`.
///
/// Flow: triggered post-onboarding → notification authorization prompt →
/// APNs registration → BettyAppDelegate forwards the APNs token to FCM →
/// FCM issues an FCM registration token → MessagingDelegate fires →
/// token delivered to `handleFCMToken(_:)` → POSTed once per distinct token.
/// Registration failures (simulators without push support, missing
/// `aps-environment` entitlement) degrade silently to `.unavailable`.
@Observable
final class PushRegistrationService: NSObject, MessagingDelegate {
    enum Phase: Equatable {
        case idle
        case denied
        case awaitingToken
        case registered(token: String)
        /// APNs registration failed (e.g. simulator) — graceful no-op.
        case unavailable
    }

    static let sentTokenDefaultsKey = "betty:push-token-sent"

    private(set) var phase: Phase = .idle

    private let defaults: UserDefaults
    private let sendToken: @MainActor (String) async throws -> Void
    private let requestAuthorization: @MainActor () async -> Bool
    private let registerWithAPNs: @MainActor () -> Void
    private let fetchFCMToken: () async -> String?

    init(defaults: UserDefaults = .standard,
         sendToken: @escaping @MainActor (String) async throws -> Void,
         requestAuthorization: @escaping @MainActor () async -> Bool = PushRegistrationService.systemRequestAuthorization,
         registerWithAPNs: @escaping @MainActor () -> Void = { UIApplication.shared.registerForRemoteNotifications() },
         fetchFCMToken: @escaping () async -> String? = PushRegistrationService.systemFetchFCMToken) {
        self.defaults = defaults
        self.sendToken = sendToken
        self.requestAuthorization = requestAuthorization
        self.registerWithAPNs = registerWithAPNs
        self.fetchFCMToken = fetchFCMToken
        super.init()
        // Guard against _FIRMessagingExceptionPlatformNotConfigured: if
        // FirebaseApp.configure() was skipped (plist absent — the documented
        // graceful-degrade path), Messaging.messaging() raises an exception.
        // Only set the delegate when Firebase is actually configured.
        if FirebaseApp.app() != nil {
            Messaging.messaging().delegate = self
        }
    }

    /// Post-onboarding trigger. Repeat calls retry an unsent token; iOS
    /// only shows the system prompt once per install.
    func registerIfNeeded() async {
        switch phase {
        case .denied, .awaitingToken:
            return
        case .registered(let token):
            await sendIfUnsent(token)
        case .idle, .unavailable:
            guard await requestAuthorization() else {
                phase = .denied
                return
            }
            phase = .awaitingToken
            registerWithAPNs()
            // Same-install account switch: FCM won't fire the delegate again because the
            // token didn't change. Fetch the cached token explicitly so the new user's
            // token reaches the backend even when the underlying registration is stale.
            if let token = await fetchFCMToken() {
                await handleFCMToken(token)
            }
        }
    }

    /// MessagingDelegate entrypoint exposed for tests. Production code
    /// receives this via `messaging(_:didReceiveRegistrationToken:)`.
    @MainActor
    func handleFCMToken(_ token: String?) async {
        guard let token, !token.isEmpty else { return }
        phase = .registered(token: token)
        await sendIfUnsent(token)
    }

    /// `didFailToRegisterForRemoteNotificationsWithError` — expected on simulators.
    func handleRegistrationFailure(_ error: Error) {
        phase = .unavailable
    }

    /// Sign-out wipes the sent marker so the next account re-POSTs its token.
    func resetForSignOut() {
        defaults.removeObject(forKey: Self.sentTokenDefaultsKey)
        phase = .idle
    }

    // MARK: - MessagingDelegate

    nonisolated func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
        Task { @MainActor in
            await self.handleFCMToken(fcmToken)
        }
    }

    // MARK: - Helpers

    private func sendIfUnsent(_ token: String) async {
        guard defaults.string(forKey: Self.sentTokenDefaultsKey) != token else { return }
        do {
            try await sendToken(token)
            defaults.set(token, forKey: Self.sentTokenDefaultsKey)
        } catch {
            // Offline / API failure: token stays unsent; retried on the next sign-in.
        }
    }

    private static func systemRequestAuthorization() async -> Bool {
        (try? await UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .badge, .sound])) ?? false
    }

    private static func systemFetchFCMToken() async -> String? {
        // Guard: Messaging.messaging() throws if Firebase is not configured.
        guard FirebaseApp.app() != nil else { return nil }
        return await withCheckedContinuation { continuation in
            Messaging.messaging().token { token, _ in
                continuation.resume(returning: token)
            }
        }
    }
}

File: android/app/src/main/java/social/betty/core/push/PushRegistrationService.kt
package social.betty.core.push

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Persists the last token successfully POSTed (iOS `UserDefaults["betty:push-token-sent"]`). */
interface SentTokenStore {
    fun get(): String?
    fun set(token: String?)
}

/**
 * Port of iOS `PushRegistrationService`: obtains the FCM registration token and POSTs it to
 * `POST /user/me/add_push_token`, exactly once per distinct token, with retry-on-failure,
 * sign-out reset, and same-install account-switch re-send. All side effects are injected so the
 * JVM unit test can drive it without Android/Firebase.
 *
 * Phases mirror iOS: Idle → (prompt) → AwaitingToken → Registered; Denied is terminal; Unavailable
 * is the no-google-services.json state.
 */
class PushRegistrationService(
    private val appScope: CoroutineScope,
    private val sentTokenStore: SentTokenStore,
    private val sendToken: suspend (String) -> Unit,        // = api.addPushToken
    private val requestAuthorization: suspend () -> Boolean, // POST_NOTIFICATIONS grant state
    private val fetchToken: suspend () -> String?,           // guarded FirebaseMessaging token
    private val isFirebaseConfigured: () -> Boolean,         // false on the no-google-services.json build
) {
    sealed interface Phase {
        data object Idle : Phase
        data object Denied : Phase
        data object AwaitingToken : Phase
        data class Registered(val token: String) : Phase
        data object Unavailable : Phase // not configured (no json) or token fetch failed — retryable
    }

    private val _phase = MutableStateFlow<Phase>(Phase.Idle)
    val phase: StateFlow<Phase> = _phase.asStateFlow()

    /**
     * Post-Ready trigger. Prompts once per install; repeat calls retry an unsent token. Same-install
     * account switch: FCM won't refire `onNewToken` (token unchanged), so we explicitly fetch the
     * cached token and POST it for the new user.
     */
    suspend fun registerIfNeeded() {
        when (val current = _phase.value) {
            Phase.Denied, Phase.AwaitingToken -> return // terminal / in-flight
            is Phase.Registered -> sendIfUnsent(current.token) // retry path
            Phase.Idle, Phase.Unavailable -> {
                // No google-services.json → push can never work. Mark Unavailable WITHOUT a
                // pointless permission prompt; a later registerIfNeeded re-checks (retryable).
                if (!isFirebaseConfigured()) {
                    _phase.value = Phase.Unavailable
                    return
                }
                if (!requestAuthorization()) {
                    _phase.value = Phase.Denied
                    return
                }
                _phase.value = Phase.AwaitingToken
                val token = fetchToken()
                if (token != null) {
                    handleToken(token)
                } else {
                    // Transient token-fetch failure → Unavailable (retryable, like iOS .unavailable)
                    // rather than stranded in the in-flight AwaitingToken state.
                    _phase.value = Phase.Unavailable
                }
            }
        }
    }

    /** Called from [BettyMessagingService.onNewToken]. */
    fun onNewToken(token: String) {
        appScope.launch { handleToken(token) }
    }

    suspend fun handleToken(token: String?) {
        if (token.isNullOrEmpty()) return
        _phase.value = Phase.Registered(token)
        sendIfUnsent(token)
    }

    /** Sign-out clears the marker so the next account re-POSTs its token. */
    fun resetForSignOut() {
        sentTokenStore.set(null)
        _phase.value = Phase.Idle
    }

    private suspend fun sendIfUnsent(token: String) {
        if (sentTokenStore.get() == token) return
        // On failure the marker stays unset, so the next registerIfNeeded() (Registered phase) retries.
        runCatching { sendToken(token) }.onSuccess { sentTokenStore.set(token) }
    }
}
